File: extract_entries.py
import xml.etree.ElementTree as ET
import math
import json
import re

def parse_svg_transform(transform_str):
    """Simple parser for rotate and scale transforms."""
    transforms = []
    if not transform_str:
        return transforms
    
    # Matches rotate(angle, [cx, cy]) or scale(sx, [sy]) or matrix(...)
    # We only handle rotate(-90) and scale(-1) as observed
    
    # Check for rotate
    rotate_match = re.search(r'rotate\(([^)]+)\)', transform_str)
    if rotate_match:
        args = [float(x) for x in rotate_match.group(1).replace(',', ' ').split()]
        transforms.append(('rotate', args))
        
    # Check for scale
    scale_match = re.search(r'scale\(([^)]+)\)', transform_str)
    if scale_match:
        args = [float(x) for x in scale_match.group(1).replace(',', ' ').split()]
        transforms.append(('scale', args))
        
    # Check for translate
    translate_match = re.search(r'translate\(([^)]+)\)', transform_str)
    if translate_match:
        args = [float(x) for x in translate_match.group(1).replace(',', ' ').split()]
        transforms.append(('translate', args))

    # Check for matrix
    matrix_match = re.search(r'matrix\(([^)]+)\)', transform_str)
    if matrix_match:
        args = [float(x) for x in matrix_match.group(1).replace(',', ' ').split()]
        transforms.append(('matrix', args))
        
    return transforms

def apply_transforms(x, y, w, h, transforms):
    # Calculate initial center
    cx = x + w / 2
    cy = y + h / 2
    
    for t_type, args in transforms:
        if t_type == 'rotate':
            angle = args[0]
            # Assumes rotation around (0,0) if no cx, cy provided
            # Standard SVG rotation
            rad = math.radians(angle)
            cos_a = math.cos(rad)
            sin_a = math.sin(rad)
            
            ncx = cx * cos_a - cy * sin_a
            ncy = cx * sin_a + cy * cos_a
            cx, cy = ncx, ncy
            
        elif t_type == 'scale':
            sx = args[0]
            sy = args[1] if len(args) > 1 else sx
            cx = cx * sx
            cy = cy * sy
            
        elif t_type == 'translate':
            tx = args[0]
            ty = args[1] if len(args) > 1 else 0
            cx = cx + tx
            cy = cy + ty
            
        elif t_type == 'matrix':
            # matrix(a b c d e f)
            # x' = ax + cy + e
            # y' = bx + dy + f
            if len(args) == 6:
                a, b, c, d, e, f = args
                ncx = a * cx + c * cy + e
                ncy = b * cx + d * cy + f
                cx, cy = ncx, ncy

    return cx, cy

def extract_entries(svg_path, floor_id):
    tree = ET.parse(svg_path)
    root = tree.getroot()
    
    # Namespaces
    ns = {
        'svg': 'http://www.w3.org/2000/svg',
        'inkscape': 'http://www.inkscape.org/namespaces/inkscape',
        'xlink': 'http://www.w3.org/1999/xlink'
    }
    
    entries = []
    
    # Find all elements with inkscape:label starting with "Wejscie"
    # We iterate all elements because they might be nested
    for elem in root.iter():
        label = elem.get('{http://www.inkscape.org/namespaces/inkscape}label')
        if label and label.lower().startswith('wejscie'):
            # It's an entry point
            x = float(elem.get('x', 0))
            y = float(elem.get('y', 0))
            width = float(elem.get('width', 0))
            height = float(elem.get('height', 0))
            transform = elem.get('transform')
            
            transforms = parse_svg_transform(transform)
            
            # Calculate visual center
            final_x, final_y = apply_transforms(x, y, width, height, transforms)
            
            entries.append({
                'floor_id': floor_id,
                'label': label,
                'room_name': label.replace('Wejscie_', '').replace('wejscie_', ''),
                'x': final_x,
                'y': final_y
            })
            
    return entries

floor_0_entries = extract_entries(r'assets\maps\floor_0.svg', 'FLOOR_0')
floor_1_entries = extract_entries(r'assets\maps\floor_1.svg', 'FLOOR_1')

all_entries = floor_0_entries + floor_1_entries

print(json.dumps(all_entries, indent=4))

File: extract_entries.js
const fs = require('fs');
const path = require('path');

function parseTransform(transformStr) {
    const transforms = [];
    if (!transformStr) return transforms;

    const rotateMatch = transformStr.match(/rotate\(([^)]+)\)/);
    if (rotateMatch) {
        const args = rotateMatch[1].replace(/,/g, ' ').trim().split(/\s+/).map(Number);
        transforms.push({ type: 'rotate', args });
    }

    const scaleMatch = transformStr.match(/scale\(([^)]+)\)/);
    if (scaleMatch) {
        const args = scaleMatch[1].replace(/,/g, ' ').trim().split(/\s+/).map(Number);
        transforms.push({ type: 'scale', args });
    }

    // Matrix handling if needed
    const matrixMatch = transformStr.match(/matrix\(([^)]+)\)/);
    if (matrixMatch) {
        const args = matrixMatch[1].replace(/,/g, ' ').trim().split(/\s+/).map(Number);
        transforms.push({ type: 'matrix', args });
    }

    return transforms;
}

function applyTransforms(x, y, w, h, transforms) {
    // Initial center
    let cx = x + w / 2;
    let cy = y + h / 2;

    for (const t of transforms) {
        if (t.type === 'rotate') {
            const angle = t.args[0];
            const rad = angle * Math.PI / 180;
            const cos = Math.cos(rad);
            const sin = Math.sin(rad);

            // Rotation around (0,0) standard
            const ncx = cx * cos - cy * sin;
            const ncy = cx * sin + cy * cos;
            cx = ncx;
            cy = ncy;
        } else if (t.type === 'scale') {
            const sx = t.args[0];
            const sy = t.args.length > 1 ? t.args[1] : sx;
            cx *= sx;
            cy *= sy;
        } else if (t.type === 'matrix') {
            // matrix(a b c d e f)
            // x' = ax + cy + e
            // y' = bx + dy + f
            if (t.args.length === 6) {
                const [a, b, c, d, e, f] = t.args;
                const ncx = a * cx + c * cy + e;
                const ncy = b * cx + d * cy + f;
                cx = ncx;
                cy = ncy;
            }
        }
    }
    return { x: cx, y: cy };
}

function extractEntries(filePath, floorId) {
    const content = fs.readFileSync(filePath, 'utf8');
    const entries = [];

    // Regex filtering for image tags with Wejscie label. 
    // We capture the whole tag content
    const tagRegex = /<image([^>]+)inkscape:label="Wejscie([^"]+)"([^>]*)\/>/gi;
    // Actually the label might be before or after other attributes, so let's match generic image tag first

    // Split by <image to be robust against multiline
    const chunks = content.split('<image');

    chunks.forEach(chunk => {
        // We only care about the part before the closing /> or >
        const endIdx = chunk.indexOf('/>');
        if (endIdx === -1) return; // Simplified: assumes self-closing for images which is standard

        const tagContent = chunk.substring(0, endIdx);

        if (tagContent.includes('inkscape:label="Wejscie')) {
            // Check label
            const labelMatch = tagContent.match(/inkscape:label="([^"]+)"/);
            if (!labelMatch) return;
            const label = labelMatch[1];
            if (!label.toLowerCase().startsWith('wejscie')) return;

            // Extract coords
            const xMatch = tagContent.match(/\sx="([^"]+)"/);
            const yMatch = tagContent.match(/\sy="([^"]+)"/);
            const wMatch = tagContent.match(/\swidth="([^"]+)"/);
            const hMatch = tagContent.match(/\sheight="([^"]+)"/);
            const tMatch = tagContent.match(/\stransform="([^"]+)"/);

            const x = xMatch ? parseFloat(xMatch[1]) : 0;
            const y = yMatch ? parseFloat(yMatch[1]) : 0;
            const w = wMatch ? parseFloat(wMatch[1]) : 0;
            const h = hMatch ? parseFloat(hMatch[1]) : 0;
            const transform = tMatch ? tMatch[1] : '';

            const transforms = parseTransform(transform);
            const finalPos = applyTransforms(x, y, w, h, transforms);

            entries.push({
                floor_id: floorId,
                label: label,
                room_name: label.replace(/^Wejscie_/i, ''),
                x: finalPos.x,
                y: finalPos.y
            });
        }
    });

    return entries;
}

const entries0 = extractEntries('assets/maps/floor_0.svg', 'FLOOR_0');
const entries1 = extractEntries('assets/maps/floor_1.svg', 'FLOOR_1');

console.log(JSON.stringify([...entries0, ...entries1], null, 2));
